timemachine/potentials/bonded.py

```diff
@@ -147,6 +148,87 @@ def harmonic_bond(conf, params, box, lamb, bond_idxs, lamb_mult=None, lamb_offse
     return jnp.sum(energy)
 
 
+def get_cos_angle(ci, cj, ck):
+
+    ci = jnp.array(ci)
+    cj = jnp.array(cj)
+    ck = jnp.array(ck)
+
+    # compute cosine angle of three points
+    vij = ci - cj
+    vkj = ck - cj
+    top = jnp.sum(jnp.multiply(vij, vkj), -1)
+    bot = jnp.linalg.norm(vij, axis=-1) * jnp.linalg.norm(vkj, axis=-1)
+
+    tb = top / bot
+
+    return tb
+
+
+def get_centroid_cos_angles(conf, angle_idxs):
+
+    x_is = []
+    x_js = []
+    x_ks = []
+
+    for (a, b), j, k in angle_idxs:
+        x_j = conf[j]
+        # compute v-site from unit vectors to avoid bond length
+        # change differences
+        x_a, x_b = conf[a], conf[b]
+        x_ia = x_a - x_j
+        x_ia = x_j + x_ia / jnp.linalg.norm(x_ia)
+        x_ib = x_b - x_j
+        x_ib = x_j + x_ib / jnp.linalg.norm(x_ib)
+        x_i = (x_ia + x_ib) / 2.0
+        x_k = conf[k]
+        x_is.append(x_i)
+        x_js.append(x_j)
+        x_ks.append(x_k)
+
+    return get_cos_angle(x_is, x_js, x_ks)
+
+
+def harmonic_x_angle(conf, params, box, lamb, angle_idxs):
+    if len(angle_idxs) == 0:
+        return 0.0
+
+    params = jnp.array(params)
+    x_is = []
+    x_js = []
+    x_ks = []
+    for (j, a), (j, b), (j, c) in angle_idxs:
```

Comment:
Find it a bit odd that you reuse the variable `j` three times here. Presumably always the same value, but might warrant either an assertion or a comment.